### ISSUE TITLE
Possible issues with Python3.6 ?

### ISSUE THREAD

**Issue by pvyaka01** [StackStorm-Exchange/stackstorm-consul#33]
Looks like when we installed StackStorm 3.5 consul pack calls started failing. One example is consul.kv_get - it's not returning a list but a complete string which is breaking workflows. Here's the KV:
Name: test_kv
Value: {"test_dc": ["node1", "node2"]}

Example:
**Stackstorm 3.5 - output**
 st2 run consul.kv_get key=test_kv dc=dc1
.
id: 61800b15435ff20388eaa484
action.ref: consul.kv_get
context.user: ansible
parameters:
  dc: dc1
  key: test_kv
status: succeeded
start_timestamp: Mon, 01 Nov 2021 15:43:17 UTC
end_timestamp: Mon, 01 Nov 2021 15:43:18 UTC
result:
  exit_code: 0
  **result: '[''194027065'', {''LockIndex'': 0, ''Key'': ''test_kv'', ''Flags'': 0, ''Value'': b''{"test_dc": ["node1", "node2"]}'', ''CreateIndex'': 194027065, ''ModifyIndex'': 194027065}]'**
  stderr: ''
  stdout: ''

**Stackstorm 3.2.0 output**
 st2 run consul.kv_get key=test_kv dc=dc1
..
id: 61800b83fe40f5a9a027d2f9
action.ref: consul.kv_get
context.user: ansible
parameters:
  dc: dc1
  key: test_kv
status: succeeded
start_timestamp: Mon, 01 Nov 2021 15:45:07 UTC
end_timestamp: Mon, 01 Nov 2021 15:45:09 UTC
**result:
  exit_code: 0
  result:
  - '194027065'
  - CreateIndex: 194027065
    Flags: 0
    Key: test_kv
    LockIndex: 0
    ModifyIndex: 194027065
    Value: '{"test_dc": ["node1", "node2"]}'
  stderr: ''
  stdout: ''**

**Comment by nzlosh** [comment]
Since the value isn't a simple data type (i.e. an object), you'll need to set the `from_json` flag to `true` so the json object is deserialised.

**Comment by pvyaka01** [comment]
But we didn't have to do it before with 3.2.

**Comment by nzlosh** [comment]
When you upgraded to StackStorm 3.5, did you reinstall the consul pack?

**Comment by pvyaka01** [comment]
This is the current version as I understand it and we're using it.
# consul integration pack v0.6.7

Thanks!

**Comment by nzlosh** [comment]
I would like to understand how the StackStorm instance was upgraded.  Did the consul pack get installed when the St2 installation was upgraded to v3.5?

**Comment by pvyaka01** [comment]
We actually didn't upgrade, it's a fresh StackStorm install and we installed consul pack using st2 pack install

**Comment by nzlosh** [comment]
OK, thank you for clarifying that.  I suspect the underlying Python Consul module is not working correctly with Python3.6.  Unfortunately the module was abandoned in 2018.  https://github.com/criteo-forks/py-consul or https://github.com/poppyred/python-consul2/ both appear to have forked from the original projected and seen significant development.

**Comment by pvyaka01** [comment]
Ok, will try to install those and see if it fixes the issue. Thanks

**Comment by pvyaka01** [comment]
unfortunately, no luck, result is the same:
 st2 run consul.kv_get key=test_kv dc=dc1
.
id: 618185b22da0371fc699efed
action.ref: consul.kv_get
context.user: ansible
parameters:
  dc: dc1
  key: test_kv
status: succeeded
start_timestamp: Tue, 02 Nov 2021 18:38:42 UTC
end_timestamp: Tue, 02 Nov 2021 18:38:43 UTC
result:
  exit_code: 0
  result: '[''194028178'', {''LockIndex'': 0, ''Key'': ''test_kv'', ''Flags'': 0, ''Value'': b''{"test_dc": {"node1", "node2"}}'', ''CreateIndex'': 194027065, ''ModifyIndex'': 194028178}]'
  stderr: ''
  stdout: ''

**Comment by mjtice** [comment]
I think the problem is that with Python3 (ST2 3.5), the `Value` returned from `consul.kv_get` is a byte-string:

>If you are working with web service libraries such as urllib (formerly urllib2) and requests, network sockets, binary files, or serial I/O with pySerial  you will find that data is now stored as byte strings.
source: https://blog.feabhas.com/2019/02/python-3-unicode-and-byte-strings/

Modifying `kv_get.py` like so seems to do the trick:
from:
```
if from_json and not keys:
            if isinstance(res, dict):
                res["Value"] = self.from_json(res["Value"])
            if isinstance(res, list):
                for item in res:
                    item["Value"] = self.from_json(item["Value"])
```
to:
```
if from_json and not keys:
            if isinstance(res, dict):
                res["Value"] = self.from_json(res["Value"].decode("utf-8"))
            if isinstance(res, list):
                for item in res:
                    item["Value"] = self.from_json(item["Value"].decode("utf-8"))
```

Additionally, `python-consul2` seems to be the "accepted" replacement according to Hashicorp:
https://www.consul.io/api-docs/libraries-and-sdks

**Comment by mjtice** [comment]
Created PR https://github.com/StackStorm-Exchange/stackstorm-consul/pull/34

**Comment by nzlosh** [comment]
From what I understood in this particular issue the `from_json` flag isn't being used.  I think the reason the object is being returned as a string is because it's not valid JSON at all and can't be deserialised to a native python dictionary.  Either way, this issue is resolved as working as expected or fixed by #34.  I'm going to close this issue and merge the PR.  Nice job @mjtice :+1:
